### PR TITLE
[DNM]: rgw: making implicit_tenants backwards compatible

### PR DESCRIPTION
**Pull request by mdw-at-linuxbox**
In jewel, "rgw keystone implicit tenants" only applied to swift. As of
luminous), this option applies to s3 also.
Sites that used this feature with jewel now have outstanding data that
depends on the old behavior.

The fix here is to expand "rgw keystone implicit tenants" so that it
can be set to any of "none", "all", "s3" or "swift" (also 0=false=none,
1=true=all).

Fixes: http://tracker.ceph.com/issues/24348

Signed-off-by: Marcus Watts <mwatts@redhat.com>

### CODE REVIEW

File: src/common/options.cc [hunk 1]
```diff
@@ -5217,12 +5217,13 @@ std::vector<Option> get_rgw_options() {
     .set_default(true)
     .set_description("Should RGW verify the Keystone server SSL certificate."),
 
-    Option("rgw_keystone_implicit_tenants", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
-    .set_default(false)
+    Option("rgw_keystone_implicit_tenants", Option::TYPE_STR, Option::LEVEL_ADVANCED)
+    .set_default("false")
+    .set_enum_allowed( { "false", "true", "swift", "s3", "both", "0", "1", "none" } )
```
Comment (mattbenjamin) [inline src/common/options.cc]: placeholder:  these need to be well documented;  even though complex, should a terse version go in the long docu?

File: src/rgw/rgw_auth.cc [hunk 2]
```diff
@@ -420,6 +452,8 @@ void rgw::auth::RemoteApplier::load_acct_info(RGWUserInfo& user_info) const
    * that belongs to the authenticated identity. Another policy may be
    * applied by using a RGWThirdPartyAccountAuthApplier decorator. */
   const rgw_user& acct_user = info.acct_user;
+  bool split_mode, implicit_tenant;
+  std::tie(implicit_tenant, split_mode) = implicit_tenants(cct);
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_auth.cc]: I think this is a nit

[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_auth.cc]: (its easy, but maybe we should prefer std::tie just now due to need to backport)

File: src/rgw/rgw_auth.cc [hunk 3]
```diff
@@ -385,7 +385,39 @@ void rgw::auth::RemoteApplier::to_str(std::ostream& out) const
       << ", is_admin=" << info.is_admin << ")";
 }
 
+bool static inline implicit_tenants_enabled_for_both(CephContext * const cct,
+	std::string &s, bool &split_mode)
+{
+  s = cct->_conf->get_val<std::string>("rgw_keystone_implicit_tenants");
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_auth.cc]: yeah, if we look below, we see this function getting called a few times apparently in one op?

File: src/rgw/rgw_auth.h [hunk 4]
```diff
@@ -396,17 +399,18 @@ class RemoteApplier : public IdentityApplier {
   const acl_strategy_t extra_acl_strategy;
 
   const AuthInfo info;
-  const bool implicit_tenants;
+  std::tuple<bool,bool> (*implicit_tenants)(CephContext*const);
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_auth.h]: that would be slightly more readable, but the name of the variable adds some clarity already

File: src/rgw/rgw_auth.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/rgw/rgw_auth.cc]: I guess not, these are called in different authorizers

**Comment by mdw-at-linuxbox** [comment]
I've pushed a new version that I hope addresses various critiques.  Using "md_config_obs_t" turns out to make the patch quite a bit more complicated.

**Comment by mdw-at-linuxbox** [comment]
One weirdness - "S3AuthFactory" seems not to be used.  I did an "#if 0" on it temporarily (and it built fine that way).  Anyone know what this is for?

**Comment by smithfarm** [comment]
@mdw-at-linuxbox It looks like the original commit underwent some changes as a result of review. Were those changes reflected in the luminous version at https://github.com/ceph/ceph/pull/22363 (which is already merged)? If not, would it make sense to open a follow-on luminous PR to sync up the diff?

File: src/rgw/rgw_rest_s3.h [hunk 5]
```diff
@@ -888,14 +888,18 @@ class S3AnonymousEngine : public rgw::auth::AnonymousEngine {
 };
 
 
+#if 0
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.h]: ?? is this class unused? why remove and also make changes to it?

**Comment by yuriw** [comment]
[wip-yuri-testing-2018-07-04-2355](https://trello.com/c/uMrCvIHO/335-wip-yuri-testing-2018-07-04-2355)

**Comment by smithfarm** [comment]
To answer the question I posed in https://github.com/ceph/ceph/pull/22378#issuecomment-396971924 I undertook a visual comparison of this PR with its luminous "backport"  #22363

The only code difference I found was in src/rgw/rgw_rest_s3.h, inside the block that is being disabled via `#if 0` so I guess the difference doesn't actually matter.

**Comment by mattbenjamin** [comment]
@vumrao marcus says that he'd like to pull these csommits into his larger bucket mv PR, as he has made some changes

**Comment by vumrao** [comment]
@mattbenjamin thank you Matt.

**Comment by mdw-at-linuxbox** [comment]
I pushed a new version - got rid of the #if 0.  Also contains fixes raised by reviewers above, and a fix for the changing api of master's md config observer.  This should be virtually identical to the luminous patch excepting for the necessary adaptions to master.

**Pull request closed without merging**